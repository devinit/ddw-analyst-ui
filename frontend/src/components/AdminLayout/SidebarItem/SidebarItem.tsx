import * as React from 'react';
import { SidebarLink, SidebarLinkProps } from '../SidebarLink';
import classNames = require('classnames');

export interface SidebarItemProps {
  to: string;
  active?: boolean;
}

export class SidebarItem extends React.Component<SidebarItemProps> {
  static Link = SidebarLink;

  render() {
    return (
      <li className={ classNames('nav-item', { active: this.props.active }) }>
        { this.renderRoot() }
        { this.renderLinks() }
      </li>
    );
  }

  private renderRoot(): React.ReactNode {
    if (React.Children.count(this.props.children) > 1) {
      const rootLink = React.Children.map(this.props.children, child => {
        if (React.isValidElement<SidebarLinkProps>(child) && child.type === SidebarLink && child.props.root) {
          return child;
        }
      })[0];

      if (rootLink) {
        return React.cloneElement(rootLink, { to: this.props.to });
      }

      console.error('A root link is required in a sidebar item with more than one child'); // tslint:disable-line
    }

    return React.Children.map(this.props.children, child => {
      if (React.isValidElement<SidebarLinkProps>(child) && child.type === SidebarLink && child.props.root) {
        return React.cloneElement(child, { single: true });
      }
    });
  }

  private renderLinks(): React.ReactNode {
    if (React.Children.count(this.props.children) > 1) {
      return (
        <div className="collapse" id={ this.props.to }>
          <ul className="nav">
            {
              React.Children.map(this.props.children, child => {
                if (React.isValidElement<SidebarItemProps>(child) && child.type === SidebarItem) {
                  return child;
                }
              })
            }
          </ul>
        </div>
      );
    }

    return React.Children.map(this.props.children, child => {
      if (React.isValidElement<SidebarLinkProps>(child) && child.type === SidebarLink && !child.props.root) {
        return child;
      }
    });
  }
}
