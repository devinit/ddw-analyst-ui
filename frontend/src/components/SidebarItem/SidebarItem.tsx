import classNames from 'classnames';
import * as React from 'react';
import { SidebarLink, SidebarLinkProps } from '../SidebarLink';

export interface SidebarItemProps {
  collapseId?: string;
  active?: boolean;
}

export class SidebarItem extends React.Component<SidebarItemProps> {
  static Link = SidebarLink;

  render() {
    return (
      <li className={ classNames('nav-item', { active: this.props.active }) }>
        { this.renderRoot() }
        { this.renderLinks() }
      </li>
    );
  }

  private renderRoot(): React.ReactNode {
    if (React.Children.count(this.props.children) > 1) {
      const rootLink = React.Children.map(this.props.children, child => {
        if (React.isValidElement<SidebarLinkProps>(child) && child.type === SidebarLink && child.props.root) {
          return child;
        }
      })[0];

      if (rootLink) {
        return React.cloneElement(rootLink, { caret: true });
      }

      console.error('A root link is required in a sidebar item with more than one child'); // tslint:disable-line
    }

    return React.Children.map(this.props.children, child => {
      if (React.isValidElement<SidebarLinkProps>(child) && child.type === SidebarLink && child.props.root) {
        return React.cloneElement(child, { caret: false });
      }
    });
  }

  private renderLinks(): React.ReactNode {
    if (React.Children.count(this.props.children) > 1) {
      return (
        <div className="collapse" id={ this.props.collapseId }>
          <ul className="nav">
            {
              React.Children.map(this.props.children, child => {
                if (React.isValidElement<SidebarItemProps>(child) && child.type === SidebarItem) {
                  return child;
                }
              })
            }
          </ul>
        </div>
      );
    }

    return React.Children.map(this.props.children, child => {
      if (React.isValidElement<SidebarLinkProps>(child) && child.type === SidebarLink && !child.props.root) {
        return child;
      }
    });
  }
}
