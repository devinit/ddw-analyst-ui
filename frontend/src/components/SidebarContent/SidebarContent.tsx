import classNames from 'classnames';
import * as React from 'react';
import { SidebarItem, SidebarItemProps } from '../SidebarItem';

interface SidebarContentProps {
  mobile?: boolean;
  children?: React.ReactNode;
}

export class SidebarContent extends React.Component<SidebarContentProps> {
  static Item = SidebarItem;

  render() {
    return (
      <ul className={classNames('nav', { 'navbar-nav nav-mobile-menu': this.props.mobile })}>
        {React.Children.map(this.props.children, (child) => {
          if (React.isValidElement<SidebarItemProps>(child) && child.type === SidebarItem) {
            return child;
          }
        })}
      </ul>
    );
  }
}
