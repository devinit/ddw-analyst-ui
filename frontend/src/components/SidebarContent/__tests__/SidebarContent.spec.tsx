import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { SidebarContent } from '../SidebarContent';
import { SidebarItem } from '../../SidebarItem';

test('renders correctly with the default props', () => {
  const renderer = TestRenderer.create(<SidebarContent />).toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders SidebarItem components', () => {
  const renderer = TestRenderer.create(
    <SidebarContent>
      <SidebarItem />
    </SidebarContent>,
  ).toJSON();

  expect(renderer).toMatchSnapshot();
});

test('renders only SidebarItem components', () => {
  const renderer = TestRenderer.create(
    <SidebarContent>
      <div>Awesome</div>
    </SidebarContent>,
  ).toJSON();

  expect(renderer).toMatchSnapshot();
});
