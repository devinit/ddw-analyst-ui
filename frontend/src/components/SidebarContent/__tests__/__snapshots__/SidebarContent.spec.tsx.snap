// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SidebarItem components 1`] = `
<div
  className="sidebar-wrapper"
>
  <ul
    className="nav"
  >
    <li
      className="nav-item"
    />
  </ul>
</div>
`;

exports[`renders correctly with the default props 1`] = `
<div
  className="sidebar-wrapper"
>
  <ul
    className="nav"
  />
</div>
`;

exports[`renders only SidebarItem components 1`] = `
<div
  className="sidebar-wrapper"
>
  <ul
    className="nav"
  />
</div>
`;
